src/lib/shell/commands/DFUManager_esp32.h

```diff
@@ -0,0 +1,147 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "OTAUpdater_esp32.h"
+#include "platform/ESP32/ESP32Utils.h"
+#include <lib/support/TypeTraits.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeDelegate.h>
+#include <protocols/bdx/BdxMessages.h>
+#include <protocols/bdx/BdxTransferSession.h>
+#include <protocols/bdx/TransferFacilitator.h>
+
+namespace chip {
+namespace Shell {
+
+class DFUManager : public bdx::Initiator
+{
+public:
+    void SetInitialExchange(Messaging::ExchangeContext * ec) { mExchangeCtx = ec; }
+    CHIP_ERROR ApplyUpdate();
+    CHIP_ERROR DiscardUpdate();
+
+private:
+    void HandleTransferSessionOutput(bdx::TransferSession::OutputEvent & event) override;
+
+    uint8_t mDfuBuffer[1024];
+    bool mIsTransferComplete = false;
+};
+
+inline CHIP_ERROR DFUManager::ApplyUpdate()
+{
+    return DeviceLayer::Internal::ESP32Utils::MapError(OTAUpdater::GetInstance().Apply());
+}
+
+inline CHIP_ERROR DFUManager::DiscardUpdate()
+{
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
```

Comment:
Is this not similar to `OTAUpdater::Abort()`?

Reply:
This function is called when there is an `EMBER_ZCL_OTA_APPLY_UPDATE_ACTION_DISCONTINUE` action during applying the update. It is different from `OTAUpdater::Abort()`,